src/stoppable/Behavior.php

```diff
@@ -0,0 +1,107 @@
+<?php
+/**
+ * @link http://www.yiiframework.com/
+ * @copyright Copyright (c) 2008 Yii Software LLC
+ * @license http://www.yiiframework.com/license/
+ */
+
+namespace yii\queue\stoppable;
+
+use yii\caching\Cache;
+use yii\di\Instance;
+use yii\queue\ExecEvent;
+use yii\queue\Queue;
+
+/**
+ * Stoppable Behavior allows stopping scheduled jobs in a queue.
+ *
+ * It provides a [[stop()]] method to mark scheduled jobs as "stopped", that
+ * will prevent their execution.
+ *
+ * This behavior should be attached to the [[Queue]] component.
+ *
+ * @author Roman Zhuravlev <zhuravljov@gmail.com>
+ * @since 2.0.1
+ */
+class Behavior extends \yii\base\Behavior
```

Comment:
imo it should be `yii\queue\behaviors\Stoppable` instead of `yii\queue\stoppable\Behavior`, Behavior as a class name does not make sense to me.

Reply:
It will be used in a configuration by long name only. As well as closure behavior. Readability shouldn't be impaired.

```php
'queue' => [
    //...
    'as stoppable' => \yii\queue\stoppable\Behavior::class,
    'as closure' => \yii\queue\closure\Behavior::class,
],
```

I think to make separate behaviors namespace isn't good idea.

Reply:
> It will be used in a configuration by long name only

Why do you think that people uses long names in configuration? I almost always use short names with `use` statements at the top of the file.

Reply:
@rob006, I'm not sure. But this make sense for config files, which cover many class definitions from different namespaces.

Reply:
Anyway, you will resolve name conflicts using namespace aliases. DbConnection, RedisConnection etc. So you can add alias:

```php
use yii\queue\stoppable\Behavior as QueueStoppable;
```

Reply:
@zhuravljov Then you probably use too generic names for your classes. I'm using `SomeComponent`, `SomeManager`, `SomeEvent`, `SomeModule` etc. and then name conflicts are pretty rare. 
And having one block with list of used classes can be really useful, that's why I'm trying to always import all classes.

Reply:
Still, having `yii\queue\stoppable` with one class doesn't make much sense. We should either create `yii\queue\behaviors` namespace for all behaviors, or keep everything flat like `LogBehavior`.

Reply:
I think this is not a problem. Namespaces are added to resolve name conflicts.

Reply:
> Still, having yii\queue\stoppable with one class doesn't make much sense.

Now `yii\queue\stoppable` contains two classes.

> We should either create yii\queue\behaviors namespace for all behaviors

I don't sure that `\yii\queue\cli\Verbose` behavior must be moved to `\yii\queue\behaviors`.

Reply:
> Now yii\queue\stoppable contains two classes.

Making implementation more complicated to justify namespace is not the right direction...

> I don't sure that \yii\queue\cli\Verbose behavior must be moved to \yii\queue\behaviors.

I would not use `\yii\queue\cli\Verbose` as a name for behavior in the first place. This name and namespace does not suggest that this is a behavior. 
Yii uses more verbose naming which is IMO much more practical: https://github.com/yiisoft/yii2/tree/master/framework/behaviors

Reply:
> Making implementation more complicated to justify namespace is not the right direction

This is planned change.

> I would not use \yii\queue\cli\Verbose as a name for behavior in the first place. This name and namespace does not suggest that this is a behavior.

I talk no about name suffix. Every package class is grouped to namespace by its role, no by class type. And general `yii\queue\behaviors` namespace is a bad idea that breaks this.

Reply:
> I talk no about name suffix. Every package class is grouped to namespace by its role, no by class type.

This is extremely inconsistent to be honest. We have `yii\queue\serializers` for serializers, but doing the same for behaviors is wrong? So now we have at least three conventions for naming behaviors:
* `yii\queue\cli\Verbose`,
* `yii\queue\LogBehavior`,
* `yii\queue\stoppable\Behavior`.

We have `Interface` suffix for interfaces, `Event` suffix for events, but name for behavior is simply random.

Reply:
> We have yii\queue\serializers for serializers, but doing the same for behaviors is wrong?

Yes, because serializer implements only one role. It is `queue\serializers`.

Behavior is queue middleware. And behaviors can solve many different functions. `queue\stoppable\Behavior` allows to stop a job, `queue\closure\Behavior` serializes closures, `queue\LogBehavior` writes logs, etc.

> So now we have at least three conventions for naming behaviors

I renamed `Verbose` to `VerboseBehavior`, and now we have one naming rule:

```
yii\queue\<Role><Type>

yii\queue\LogBehavior: Role = queue\Log, Type = Behavior
yii\queue\closure\Behavior: Role = queue\closure\, Type = Behavior
yii\queue\stoppable\Behavior: Role = stoppable\, Type = Behavior
yii\queue\PushEvent: Role = queue\Push, Type = Event
etc.
```

Reply:
> I renamed Verbose to VerboseBehavior, and now we have single naming rule:

You clearly have two naming conventions:
1. `yii\queue\<Role><Type>`
2. `yii\queue\<role>\<Type>`

Anyway, I just hate using general names for specific implementations. It kills readability and productivity. I often use "Search by class name" feature in IDE - it is useless if you name all behaviors as "Behavior". You can't just import such class because you probably will get conflicts and `Behavior` will be just meaningless if you find it in code.

Reply:
This is one of common ways of naming:

- `yii\<mode>\Application`
- `yii\<package>\Module`
- `yii\gii\generators\<entity>\Generator`
- `Illuminate\Redis\Connections\Connection`
- `Illuminate\Database\Connection`
- `<vendor>\<package>\Client`

> I often use "Search by class name" feature in IDE - it is useless if you name all behaviors as "Behavior".

I'm not sure about all IDEs, but PHPStorm is able to search class usage considering namespace.

Reply:
This is not a pointless discussion.

In certain discussions there is no right or wrong, just different opinions.
In this particular case, the Yii2 staff could enrich the discussion with their opinions.

In my opinion, the best solution would be to maintain the naming pattern used in Yii2.
After all, it is an extension to a larger project and here is no place to innovate.

cc: @cebe @samdark